refactor: deduplicate the message sign/verify code

vasild opened the pull request:
The message signing and verifying logic was replicated in a few places
in the code. Consolidate in a newly introduced `MessageSign()` and
`MessageVerify()` and add unit tests for them.

vasild committed:
Deduplicate the message verifying code

The logic of verifying a message was duplicated in 2 places:

src/qt/signverifymessagedialog.cpp
  SignVerifyMessageDialog::on_verifyMessageButton_VM_clicked()

src/rpc/misc.cpp
  verifymessage()

with the only difference being the result handling. Move the logic into
a dedicated

src/util/message.cpp
  MessageVerify()

which returns a set of result codes, call it from the 2 places and just
handle the results differently in the callers.

vasild committed:
Deduplicate the message signing code

The logic of signing a message was duplicated in 3 places:

src/qt/signverifymessagedialog.cpp
  SignVerifyMessageDialog::on_signMessageButton_SM_clicked()

src/rpc/misc.cpp
  signmessagewithprivkey()

src/wallet/rpcwallet.cpp
  signmessage()

Move the logic into

src/util/message.cpp
  MessageSign()

and call it from all the 3 places.

vasild commented on src/util/message.cpp:
```diff
@@ -0,0 +1,80 @@
+// test/lint/lint-includes.sh:90 complains that a new Boost dependency is
+// introduced because even though boost::get() is used in many other places in
+// the code, nobody has bothered to include <boost/variant/get.hpp>.
+//#include <boost/variant/get.hpp> // For boost::get()
```
For some reason it is possible to use `boost::get()` without including the header that provides it. Probably included from somewhere else, maybe from QT stuff.

Just drop lines 15-20 or somehow silence the lint script to allow this header to be included (a new boost dependency is not added by this patch)?

vasild replied:
I dropped lines 15-20

fanquake commented:
Looks like this is going to overlap with #17557.

vasild commented:
Affairs with conflicting PRs:

* #17557 util: Refactor message hashing into a utility function
Both PRs target the same duplicated code! However this PR (#17577) has broader scope as it deduplicates more code.

* #17493 util: Forbid ambiguous repeated assignments in config file
Purely mechanical conflicts. Both PRs touch neighboring code. Either one can be merged first, the second will need to be adjusted (easy).

* #17473 refactor: Settings code cleanups
Purely mechanical conflicts. Same as above.

* #17399 validation: Templatize ValidationState instead of subclassing
Purely mechanical conflicts. Same as above.

* #16545 Implement missing error checking for ArgsManager flags
Purely mechanical conflicts. Same as above.

vasild commented:
@jkczyz What do you think about this PR and its relation to #17557? Merge both in one PR?

jkczyz commented:
> @jkczyz What do you think about this PR and its relation to #17557? Merge both in one PR?

I have #17557 based on an outstanding PR (#17399), though I suppose it didn't have to be. The former resulted from a comment in the latter. I deliberately kept the change small so it could hopefully be merged quicker. Though I wanted to add a test to clarify why the "magic" string is needed.

That said, from a design perspective, I prefer your PR as it hides the complexity around signing a message.

jkczyz committed:
Refactor message hashing into a utility function

And add unit test for it.

The purpose of using a preamble or "magic" text as part of signing and
verifying a message was not given when the code was repeated in a few
locations. Make a test showing how it is used to prevent inadvertently
signing a transaction.

vasild commented:
@jkczyz thanks for the feedback! I did cherry-pick the last two commits from https://github.com/bitcoin/bitcoin/pull/17557 into this PR under e3d85bcd5 - the hashing utility function and the unit test for it (to clarify why the "magic" string is needed).

vasild commented:
> ### Conflicts
> 
> Reviewers, this pull request conflicts with the following ones:
> 
>     * [#17593](https://drahtbot.github.io/bitcoin_core_issue_redirect/r/17593.html) (test: move more utility functions into test utility library by mzumsande)
> 
>     * [#17581](https://drahtbot.github.io/bitcoin_core_issue_redirect/r/17581.html) (refactor: Remove settings merge reverse precedence code by ryanofsky)
> 
>     * [#17580](https://drahtbot.github.io/bitcoin_core_issue_redirect/r/17580.html) (refactor: Add ALLOW_LIST flags and enforce usage in CheckArgFlags by ryanofsky)
> 
>     * [#17493](https://drahtbot.github.io/bitcoin_core_issue_redirect/r/17493.html) (util: Forbid ambiguous multiple assignments in config file by ryanofsky)
> 
>     * [#17473](https://drahtbot.github.io/bitcoin_core_issue_redirect/r/17473.html) (refactor: Settings code cleanups by ryanofsky)
> 
>     * [#17399](https://drahtbot.github.io/bitcoin_core_issue_redirect/r/17399.html) (validation: Templatize ValidationState instead of subclassing by jkczyz)
> 
>     * [#16545](https://drahtbot.github.io/bitcoin_core_issue_redirect/r/16545.html) (refactor: Implement missing error checking for ArgsManager flags by ryanofsky)

All of the above conflicts are mechanical, small and easy to resolve.

instagibbs commented:
concept ACK, though I'd probably be more pedantic in calling it "Bitcoin Signed Message Protocol" or something. Unfortunately there's no BIP to refer to.

vasild commented:
@instagibbs Thanks for the review! Where would you put the "Bitcoin Signed Message Protocol" label?

We do message signing and verification now. What about encryption and decryption? Would that be useful feature or just a bloat that is rarely used? If encryption may be added in the future then maybe just "Bitcoin Message Protocol" would be a better name, or "Bitcoin Text Message Protocol".

laanwj commented:
I don't think encryption / decryption in bitcoind is a good idea. It is best to avoid using keys for multiple purposes (e.g. use a signing key for encryption, and vice versa). Also it's scope creep: Bitcoin core is a validating full node. It should do that well, and not try to make a GPG-ish mess of it that tries to do everything with a mash-up of features but nothing particularly well…

Sjors commented:
See also @achow101's #18115

vasild commented:
> See also @achow101's #18115

@Sjors thanks for the notice!

This PR deduplicates message signing and message verification code.

I peeked at https://github.com/bitcoin/bitcoin/pull/18115/ and among other things it deduplicates the message signing code from 2 of 3 places and does not deduplicate the message verification code. Assuming it also fixes the [3rd place](https://github.com/bitcoin/bitcoin/pull/18115/#issuecomment-585122576) where message signing code is repeated, then it will achieve half of this PR.

Should https://github.com/bitcoin/bitcoin/pull/18115/ hit `master` before this PR, then I will trim this PR to only consolidate the message verification code.

vasild commented:
> I peeked at #18115 and among other things it deduplicates the message signing code from 2 of 3 places and does not deduplicate the message verification code. Assuming it also fixes the 3rd place...

That was fixed and now #18115 deduplicates all 3 message signing repetitions.

[MessageSign()](https://github.com/bitcoin/bitcoin/pull/17577/files#diff-f7b16fbe87fc15c6f15347e78210e501R24) from https://github.com/bitcoin/bitcoin/pull/17577 is equivalent to [SignMessage()](https://github.com/bitcoin/bitcoin/pull/18115/files#diff-d546634098d8003b6397c8643b74e47dR520) from https://github.com/bitcoin/bitcoin/pull/18115. The former has unit tests added in https://github.com/bitcoin/bitcoin/pull/17577.

achow101 commented on src/util/message.cpp:
```diff
@@ -14,7 +15,27 @@
+bool MessageSign(
+    const CKey& privkey,
+    const std::string& message,
+    std::string* signature)
```
This should be a reference (`std::string&`) instead of pointer.

vasild replied:
Done.

achow101 commented on src/test/util_tests.cpp:
```diff
@@ -2094,4 +2096,13 @@ BOOST_AUTO_TEST_CASE(message_verify)
+    const uint256 signature_hash = Hash(unsigned_tx.begin(), unsigned_tx.end());
+
+    const uint256 message_hash = MessageHash(unsigned_tx);
+    BOOST_CHECK_NE(message_hash, signature_hash);
```
I think you should have on additional check here with the full hashed message. i.e. also hash `Bitcoin Signed Message:\n...` and compare that with `message_hash`.

vasild replied:
Done.

vasild replied:
Notice: assuming `CHashWriter hasher(SER_GETHASH, 0);`, then:
```
hasher << "ab" << "cd";
```
would produce a different result than:
```
hasher << "abcd";
```

achow101 commented on src/util/message.cpp:
```diff
@@ -14,7 +15,27 @@
 #include <vector>
 
-const std::string strMessageMagic = "Bitcoin Signed Message:\n";
+static const std::string MESSAGE_MAGIC = "Bitcoin Signed Message:\n";
```
Renaming this should probably go in the next commit?

vasild replied:
Done.

achow101 reviewed (commented):
Couple of nits. otherwise ACK 1a00fb84171666451cb5a21ba454755dab6e0328

vasild commented on src/qt/signverifymessagedialog.cpp:
```diff
@@ -189,51 +189,55 @@ void SignVerifyMessageDialog::on_addressBookButton_VM_clicked()
+
+    switch (MessageVerify(address, signature, message)) {
+        case MessageVerificationResult::ERR_INVALID_ADDRESS:
+            ui->statusLabel_VM->setStyleSheet("QLabel { color: red; }");
```
Done.

Sjors commented:
tACK ad39c8fd8f2f7752e4f4508ab880873008c7452e, thanks for adding tests as well

achow101 commented:
I'm seeing a new compiler warning:
```
In file included from ./script/signingprovider.h:13,
                 from ./outputtype.h:10,
                 from rpc/misc.cpp:8:
rpc/misc.cpp: In function ‘UniValue verifymessage(const JSONRPCRequest&)’:
./sync.h:179:104: warning: control reaches end of non-void function [-Wreturn-type]
  179 | #define LOCK(cs) DebugLock<decltype(cs)> PASTE2(criticalblock, __COUNTER__)(cs, #cs, __FILE__, __LINE__)
      |                                                                                                        ^
rpc/misc.cpp:273:5: note: in expansion of macro ‘LOCK’
  273 |     LOCK(cs_main);
      |     ^~~~
```

achow101 commented on src/rpc/misc.cpp:
```diff
@@ -276,31 +276,19 @@ static UniValue verifymessage(const JSONRPCRequest& request)
+            return false;
+        case MessageVerificationResult::OK:
+            return true;
     }
```
You need another `return false` statement after this `switch`.

Sjors replied:
Nice catch. I'm not getting that `-Wreturn-type` warning on macOS. Only one of the Travis machines noticed it: https://travis-ci.org/bitcoin/bitcoin/jobs/649943882?utm_medium=notification&utm_source=github_status

vasild replied:
Added `return false;`, thanks!

Sjors commented:
re-ACK e193a84fb28068e38d5f54fbfd6208428c5bb655

achow101 commented:
ACK e193a84fb28068e38d5f54fbfd6208428c5bb655

Sjors commented:
@meshcollider some of this touches the wallet - and #18115 that builds on top of this definitely does - so maybe you can take a look?

instagibbs commented on src/util/message.h:
```diff
@@ -0,0 +1,49 @@
+ * @param[in] signature The signature in base64 format.
+ * @param[in] message The message that was signed.
+ * @return result code */
+MessageVerificationResult MessageVerify(
```
nit: I'd prefer actual types used instead of strings for all args.

instagibbs replied:
this is a non-blocking comment :) let's get this merged

instagibbs reviewed (commented):
utACK https://github.com/bitcoin/bitcoin/pull/17577/commits/e193a84fb28068e38d5f54fbfd6208428c5bb655

meshcollider reviewed (commented):
utACK e193a84fb28068e38d5f54fbfd6208428c5bb655

Nice cleanup, thanks!
